### PR TITLE
New script to run Node-DC-EIS workload 

### PR DESCRIPTION
**Pull request by uttampawar**
Hi,
I've developed the required script run_node-dc-eis.sh script (based on run_acmeair.sh) to run Node-DC-EIS on nodejsbenchmarking setup along with other workloads.
Assuming workload is installed in /home/ios/node-dc-eis/ directory (cloned from https://github.com/Node-DC/Node-DC-EIS.git) then this script can be run as follows,

$ ./run_node-dc-eis.sh /home/ios/node-dc-eis/Node-DC-EIS/

@mhdawson I've tested this on my local setup, but please let me know how I can test it on the benchmarking machine and anything else I need to do to move this forward. The way I see it, Node-DC-EIS needs to be cloned on server, who would do that?

Also, this addresses https://github.com/nodejs/benchmarking/issues/86

### CODE REVIEW

**Comment by uttampawar** [comment]
@mhdawson Michael, I'm ready to pull in Node-DC-EIS workload with submitted script.

File: experimental/benchmarks/node-dc-eis/kill_node_linux [hunk 1]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
```
Comment (kunalspathak) [inline experimental/benchmarks/node-dc-eis/kill_node_linux]: not related to this PR, but we should have common scripts in single place instead of having it every benchmark folder.

Comment (uttampawar) [inline experimental/benchmarks/node-dc-eis/kill_node_linux]: @kunalspathak Once we get Node-DC-EIS in, I can work to get all the scripts in one place or in fact we can consolidate them (at least external application specific scripts).

File: experimental/benchmarks/node-dc-eis/run_node-dc-eis.sh [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+if [ -z "$NODE" ]; then
+  echo "ERROR: Could not find a 'node' executable. Please set the NODE environment variable or update the PATH."
+  echo "node is not here: $NODE"
+  exit 1
+fi
+
+echo -e "NODE VERSION:"
+$NODE --version
+# build command
+CMD="$NODE_AFFINITY ${NODE} ${NODE_FILE}"
```
Comment (uttampawar) [inline experimental/benchmarks/node-dc-eis/run_node-dc-eis.sh]: @kunalspathak server-cluster.js is part of Node-DC-EIS codebase which is not part of this PR. To use Node-DC-EIS, like AcmeAir, someone with access needs to install the workload from https://github.com/Node-DC/Node-DC-EIS.git into /home/ios/node-dc-eis directory.

File: experimental/benchmarks/node-dc-eis/run_node-dc-eis.sh [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+REL_DIR=$(dirname $0)
+ROOT_DIR=`cd "${REL_DIR}/.."; pwd`
+echo "ROOT_DIR=${ROOT_DIR}"
+SCRIPT_DIR=${ROOT_DIR}/node-dc-eis
+ACMEAIR_DIR=${RESOURCE_DIR}/node-dc-eis
+NODEDCEIS_DIR=${RESOURCE_DIR}/Node-DC-EIS-cluster
+MONGO_DIR=${SCRIPT_DIR}
+MONGODB_PORT=28000
+
+export DB_URL="mongodb://127.0.0.1:${MONGODB_PORT}/node-dc-eis-db"
```
Comment (uttampawar) [inline experimental/benchmarks/node-dc-eis/run_node-dc-eis.sh]: ACMEAIR_DIR variable is not needed should be remove. I'll update the PR.

File: experimental/benchmarks/node-dc-eis/run_node-dc-eis.sh [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+optional RESULTSDIR ${ROOT_DIR}/results
+optional TIMEOUT 600
+RESULTSLOG=$TEST_NAME.log
+SUMLOG=score_summary.txt
+
+optional DRIVERHOST
+optional NODE_FILE server-cluster
+optional CLUSTER_MODE false
+optional PORT 9000
+optional DRIVERCMD ${RESOURCE_DIR}/Node-DC-EIS-client/runspec.py
```
Comment (uttampawar) [inline experimental/benchmarks/node-dc-eis/run_node-dc-eis.sh]: runspec.py is part of Node-DC-EIS code base. Once you clone/install the workload in appropriate directory (/home/ios/node-dc-eis) then these scripts should work fine.

File: experimental/benchmarks/node-dc-eis/run_node-dc-eis.sh [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+#shopt -s -o nounset
+
+start=`date +%s`
+#set locations so we don't end up with lots of hard coded bits throughout script
+RESOURCE_DIR=$1
+REL_DIR=$(dirname $0)
+ROOT_DIR=`cd "${REL_DIR}/.."; pwd`
+echo "ROOT_DIR=${ROOT_DIR}"
+SCRIPT_DIR=${ROOT_DIR}/node-dc-eis
+ACMEAIR_DIR=${RESOURCE_DIR}/node-dc-eis
```
Comment (uttampawar) [inline experimental/benchmarks/node-dc-eis/run_node-dc-eis.sh]: This variable is not used. I'll update the PR.

File: .gitignore [hunk 6]
```diff
@@ -1 +1,5 @@
 node_modules/
+jmeter.log
```
Comment (uttampawar) [inline .gitignore]: They get added if you want to update the codebase and submit a PR unless you include them in .gitignore.

Comment (uttampawar) [inline .gitignore]: @kunalspathak All the client and server code are part of whole package Node-DC-EIS. So that means there are few ways we can get access to the client. 
1) git clone Node-DC-EIS on the benchmarking machine once in /home/ios/node-dc-eis directory (similar to AcmeAir install), or
2) I submit the PR with whole Node-DC-EIS codebase, or
3) I can upgrade current script "run_node-dc-eis.sh" script to clone it as part of the execution.

To keep things simple, my preference is to clone the current Node-DC-EIS on the benchmarking machine until there is any update to the workload (new use case). I don't see AcmeAir codebase in the benchmarking repository that means AcmeAir workload is installed separately on the benchmarking machine and only top level scripts are developed and submitted for it. Does it make sense?

But what you (community) prefers?

Comment (kunalspathak) [inline .gitignore]: makes sense. To try out, you can work with @mhdawson  and setup the repo on benchmarking machine and run it using test job that @mhdawson created.

File: experimental/benchmarks/node-dc-eis/run_node-dc-eis.sh (hunk 2 above)
Comment (kunalspathak) [inline experimental/benchmarks/node-dc-eis/run_node-dc-eis.sh]: perfect

Comment (kunalspathak) [inline experimental/benchmarks/node-dc-eis/run_node-dc-eis.sh]: I see that you use cache to save employee information rather than going to mongodb. Is there any reason why you set a cap on maxsize/maxAge [here](https://github.com/Node-DC/Node-DC-EIS/blob/master/Node-DC-EIS-cluster/config/configuration.js#L30-L31)? Do you see memory growth if you set maxSize to default i.e. Infinite? I think the goal should be to use as much cache as you can rather than relying on mongodb because that adds other factors to the benchmark score.

Comment (uttampawar) [inline experimental/benchmarks/node-dc-eis/run_node-dc-eis.sh]: cache use is off by default but it's been added so that people can try to understand the impact of these modules/functionalities. I've not seen memory growth with this workload with or without cache mechanism.

Comment (kunalspathak) [inline experimental/benchmarks/node-dc-eis/run_node-dc-eis.sh]: Did you measure how much variance you get with and without using cache? I prefer using cache so that way we don't have any dependency on mongo database once cache is populated. Is there any reason you prefer without cache mechanism?

Comment (uttampawar) [inline experimental/benchmarks/node-dc-eis/run_node-dc-eis.sh]: @kunalspathak In my experience for benchmarking I would prefer to disable cache because that  makes network bandwidth usage important while talking to DB server. That means network bandwidth is shared between all three components, client to server to DB server which is a good way to stress the whole system not just CPU.
Current implementation shows simple way to implement it in an application, useful for end user.

**Comment by mhdawson** [comment]
@kunalspathak lets set a time next week (week of the 1st) were we can set up a CI job that tests your script and does any other checkouts etc. that are required.

**Comment by mhdawson** [comment]
Looking at some of the comments, we need to figure out how to avoid a pre-install. It is fine if we have to clone additional external repos, but we should avoid pre-config unless there is no other option.  Ideally the job will

1) build Node (we already do that)
2) clone the benchmarking repo (we already do that)
3) clone additional repos, or otherwise get artifacts needed to run the benchmarks 

3) is what we'll need to add to in order to make sure we have the required components for Node-DC (at least those that cannot just be committed into the benchmarking repo which is already being cloned.

**Comment by kunalspathak** [comment]
Sounds good @mhdawson

**Comment by mhdawson** [comment]
@uttampawar hopefully we can talk about it in the meeting today and setup a time.

**Comment by uttampawar** [comment]
@mhdawson Mike, I messed up my calendar entry again and missed the meeting on 1st Aug. Is there a way us connecting again to try test node-dc workload?

**Comment by kunalspathak** [comment]
@uttampawar , let us setup something early next week. @mhdawson - I think I have access to Jenkins, so if you want, you can skip it.

**Comment by uttampawar** [comment]
@kunalspathak I really appreciate your help. Let me know the day and time we can try this out.

**Comment by kunalspathak** [comment]
@uttampawar - I have created the doodle for it. Please let me know your availability.
@mhdawson  - We might need you as well because i realized that I don't have access to the machine. (Not sure if i will need or not though for the dry run).

https://doodle.com/poll/3c8gu3g6cp2ne4kg

**Comment by mhdawson** [comment]
I'm on holiday next week so I can't make the times suggested, but don't wait on me, would be good to get started and make as much progress as the 2 of you can.  When I'm back I can help resolve anything that needed machine access or more context of what we've done earlier.

**Comment by kunalspathak** [comment]
@uttampawar  - What is your availability? Looks like you didn't fill out the doodle poll.

**Comment by uttampawar** [comment]
@kunalspathak I'm available today (14th Aug).

**Comment by kunalspathak** [comment]
I have few meetings today. Will 8/16, 11am work for you?

**Comment by uttampawar** [comment]
@kunalspathak Yes 8/16 at 11am works for me.

**Comment by kunalspathak** [comment]
@uttampawar  - I just checked my calendar and unfortunately i didn't notice that i have few things to take care on 8/16. Let us try to meet on 8/17 instead at same time. I will share the hangout link 5 mins before the meeting.

**Comment by uttampawar** [comment]
@kunalspathak For me, 8/17 works as well.

**Comment by kunalspathak** [comment]
@mhdawson is interested in joining and he will be available on 8/17 11:30 AM PST.

**Comment by mhdawson** [comment]
I now have a potential conflict at @11:30 AM PST so 12 AM PST would be better. But if that does not work for everybody else please just go ahead and I'll make whatever part of the meeting I can.  

If its at 11 I won't make it at all, if its at 11:30 I might be 10-20 mins late and if its at 12 I should be able to make it on time.

**Comment by kunalspathak** [comment]
Ok, lets plan to meet at 12 then.

**Comment by uttampawar** [comment]
@kunalspathak Do you have calling information?

**Comment by kunalspathak** [comment]
@uttampawar, I will update https://github.com/nodejs/benchmarking/issues/131 with hangout link few minutes before the meeting.

**Comment by uttampawar** [comment]
@kunalspathak @mhdawson Does benchmarking machine has 'npm' binary installed? Since 'npm install' is the first thing we need to do before starting the workload. It seems to be the problem running node-dc-eis successfully.

**Comment by kunalspathak** [comment]
If I see the `benchmark-footprint-8.x` job, I do see `npm` being used in following way. Try doing that. 
```cmd
$NODE_ROOT/deps/npm/bin/npm-cli.js install benchmark
```

**Comment by mhdawson** [comment]
@kunalspathak can you take another look and comment on if you think its ready to land.

**Comment by kunalspathak** [comment]
Sure, will take a look sometime today.

File: experimental/benchmarks/node-dc-eis/run-dc-eis.sh [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+( sleep $TIMEOUT; echo "TIMEOUT (${TIMEOUT}s)"; echo "fail" >> $DONEFILE_TEMP; ) &
+TIMEOUT_PID=$!
+TIMEOUT_CHILD=`pgrep -P $TIMEOUT_PID`
+export OTHERPID_LIST="$OTHERPID_LIST $TIMEOUT_CHILD $TIMEOUT_PID $LOOKFORDONE_PID"
+
+LOGDIR_PREFIX=$PRODUCT/$DATE/$CUR_DATE
+SUMFILE=$LOGDIR_TEMP/$LOGDIR_PREFIX/$SUMLOG
+STDOUT_SERVER=$LOGDIR_TEMP/$LOGDIR_PREFIX/server.out
+STDOUT_CLIENT=$LOGDIR_TEMP/$LOGDIR_PREFIX/client.out
+#STDOUT_RESULTS=$LOGDIR_TEMP/$LOGDIR_PREFIX/jmeter.log #uttam. TODO Do we need this for node-dc-eis
```
Comment (kunalspathak) [inline experimental/benchmarks/node-dc-eis/run-dc-eis.sh]: > #uttam. TODO Do we need this for node-dc-eis [](start = 54, length = 45)

can you remove this?

File: experimental/benchmarks/node-dc-eis/fp.sh [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  PIDS=$1
+fi
+export PIDS
+FP_LOG=$LOGDIR_TEMP/footprint_output.log
+for SERVER_PID in $PIDS
+do
+  FOOTPRINT=`ps -p ${SERVER_PID} -o rss,vsz,comm,pid | tail -n 1 | awk '{ print $1 }'`
+  HPTOTAL=`cat /proc/meminfo | grep ^HugePages | sed 's/HugePages.*: *//g' | head -n 1`
+  HPFREE=` cat /proc/meminfo | grep ^HugePages | sed 's/HugePages.*: *//g' | head -n 2 | tail -n 1`	
+  HPRESVD=` cat /proc/meminfo | grep ^HugePages | sed 's/HugePages.*: *//g' | head -n 3 | tail -n 1`
```
Comment (kunalspathak) [inline experimental/benchmarks/node-dc-eis/fp.sh]: can you fix the formatting and have it sync with other files?

**Comment by kunalspathak** [comment]
LGTM with minor comments.

**Comment by uttampawar** [comment]
@kunalspathak @mhdawson Fixed formatting issues and added chart template for the graphs. Can this land now?

File: experimental/benchmarks/node-dc-eis/run-dc-eis.sh [hunk 9]
```diff
@@ -16,9 +16,9 @@ MONGODB_PORT=27017
 export DB_URL="mongodb://127.0.0.1:${MONGODB_PORT}/node-dc-eis-db"
 
 #these may need changing when we find out more about the machine we're running on
-NODE_AFFINITY="numactl --physcpubind=0,3"
-MONGO_AFFINITY="numactl --physcpubind=1,2"
-RUNSPEC_AFFINITY="numactl --physcpubind=1,2"
+NODE_AFFINITY="numactl --physcpubind=0,4"
```
Comment (kunalspathak) [inline experimental/benchmarks/node-dc-eis/run-dc-eis.sh]: why this change in the commit?

File: experimental/benchmarks/node-dc-eis/fp.sh [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+if [ -z $1 ]; then
+	PIDS=`ps -ef|grep /node|grep -v grep|grep -v tee|grep -v sh|awk {'print $2'}`
+else
+	PIDS=$1
+fi
+export PIDS
+FP_LOG=$LOGDIR_TEMP/footprint_output.log
+for SERVER_PID in $PIDS
+do
+				FOOTPRINT=`ps -p ${SERVER_PID} -o rss,vsz,comm,pid | tail -n 1 | awk '{ print $1 }'`
```
Comment (kunalspathak) [inline experimental/benchmarks/node-dc-eis/fp.sh]: sorry for nit-pick, but formatting of this file is not same as other files.

Comment (uttampawar) [reply experimental/benchmarks/node-dc-eis/fp.sh]: @kunalspathak I compared this to fp.sh in acmeair. It is same except one new line at the end. Which file are you comparing against?

Comment (kunalspathak) [reply experimental/benchmarks/node-dc-eis/fp.sh]: I was just comparing against the other bash files but if the formatting is similar to what we have in acmeair, then ignore my comment.

Comment (uttampawar) [reply experimental/benchmarks/node-dc-eis/fp.sh]: @kunalspathak I'd changed it for better by replacing 4-tabs by just one but then I'd reversed changes to match acmeair's fp.sh script. Anyway, is this ready for landing?

File: experimental/benchmarks/node-dc-eis/run-dc-eis.sh [hunk 11]
```diff
@@ -16,9 +16,9 @@ MONGODB_PORT=27017
...
+MONGO_AFFINITY="numactl --physcpubind=1,5"
+RUNSPEC_AFFINITY="numactl --physcpubind=2,6"
```
Comment (mhdawson) [inline experimental/benchmarks/node-dc-eis/run-dc-eis.sh]: Same here:

I agree I don't think this is correct since I believe there are only 4 CPUs.

File: experimental/benchmarks/node-dc-eis/run-dc-eis.sh (hunk 9 above)
Comment (mhdawson) [reply experimental/benchmarks/node-dc-eis/run-dc-eis.sh]: I agree I don't think this is correct since I believe there are only 4 CPUs.

File: experimental/benchmarks/node-dc-eis/run-dc-eis.sh [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+NODEDCEIS_DIR=${WORKLOAD_DIR}/Node-DC-EIS-cluster
+MONGO_DIR=${RESOURCE_DIR}/mongo3
+MONGODB_PORT=27017
+
+export DB_URL="mongodb://127.0.0.1:${MONGODB_PORT}/node-dc-eis-db"
+
+#these may need changing when we find out more about the machine we're running on
+NODE_AFFINITY="numactl --physcpubind=0,4"
+MONGO_AFFINITY="numactl --physcpubind=1,5"
+RUNSPEC_AFFINITY="numactl --physcpubind=2,6"
```
Comment (mhdawson) [inline experimental/benchmarks/node-dc-eis/run-dc-eis.sh]: Do these make sense since we only have 4 cpus?

File: experimental/benchmarks/node-dc-eis/run-dc-eis.sh (hunk 11 above)
Comment (mhdawson) [reply experimental/benchmarks/node-dc-eis/run-dc-eis.sh]: I guess its ok since it matches what is there for ACME air.  Will go ahead an land since I believe comments for @kunalspathak have been addressed.

**Comment by gareth-ellis** [comment]
Using affinities for cpus that don't exist will result in the command not being issued, and thus the run failing.

@mhdawson can you run numactl --hardware on the benchmarking machine to confirm what is there?

**Comment by mhdawson** [comment]
Look here: https://github.com/nodejs/benchmarking/blob/master/experimental/benchmarks/acmeair/run_acmeair.sh

It seems to be the same.

I'll run numactl once I finish off other things I'm working on.

**Comment by mhdawson** [comment]
This is the output
```sh
root@nodejsb1:~# numactl --hardware
available: 1 nodes (0)
node 0 cpus: 0 1 2 3 4 5 6 7
node 0 size: 7957 MB
node 0 free: 2821 MB
node distances:
node   0 
  0:  10 
```

so likely hyperthreading is on.

**Comment by mhdawson** [comment]
Landed as fcc3d8b5ac06c2adfd02bd240bd72982e66d25a4
